graylog2-server/src/main/java/org/graylog2/migrations/V20251021083100_CreatePipelineMetadata.java

```diff
@@ -0,0 +1,114 @@
+import org.graylog2.database.MongoCollection;
+import org.graylog2.database.MongoCollections;
+import org.graylog2.database.MongoConnection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.graylog.plugins.pipelineprocessor.db.mongodb.MongoDbPipelineMetadataService.INPUTS_COLLECTION_NAME;
+import static org.graylog.plugins.pipelineprocessor.db.mongodb.MongoDbPipelineMetadataService.RULES_COLLECTION_NAME;
+
+/**
+ * Migration to create the pipeline metadata collections, if any of them do not exist yet.
+ */
+public class V20251021083100_CreatePipelineMetadata extends Migration {
+    private static final Logger LOG = LoggerFactory.getLogger(V20251021083100_CreatePipelineMetadata.class);
+    private final MongoDatabase db;
+    private final PipelineResolver pipelineResolver;
+    private final PipelineAnalyzer pipelineAnalyzer;
+    private final MongoCollection<PipelineRulesMetadataDao> rulesCollection;
+    private final MongoCollection<PipelineInputsMetadataDao> inputsCollection;
+
+    @Inject
+    public V20251021083100_CreatePipelineMetadata(MongoConnection mongoConnection,
+                                                  MongoCollections mongoCollections,
+                                                  MongoDbRuleService ruleService,
+                                                  PipelineService pipelineService,
+                                                  PipelineAnalyzer pipelineAnalyzer,
+                                                  PipelineStreamConnectionsService pipelineStreamConnectionsService,
+                                                  PipelineRuleParser pipelineRuleParser,
+                                                  PipelineResolver.Factory pipelineResolverFactory) {
+        this.db = mongoConnection.getMongoDatabase();
+        this.rulesCollection = mongoCollections.collection(RULES_COLLECTION_NAME, PipelineRulesMetadataDao.class);
+        this.inputsCollection = mongoCollections.collection(INPUTS_COLLECTION_NAME, PipelineInputsMetadataDao.class);
+        this.pipelineAnalyzer = pipelineAnalyzer;
+        this.pipelineResolver = pipelineResolverFactory.create(
+                PipelineResolverConfig.of(
+                        () -> ruleService.loadAll().stream(),
+                        () -> pipelineService.loadAll().stream(),
+                        () -> pipelineStreamConnectionsService.loadAll().stream()
+                ),
+                pipelineRuleParser
+        );
+    }
+
+    @Override
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.parse("2025-10-21T08:31:00Z");
+    }
+
+    @Override
+    public void upgrade() {
+        if (collectionNotExists(RULES_COLLECTION_NAME) || collectionNotExists(INPUTS_COLLECTION_NAME)) {
```

Comment:
Nitpick: We don't need to iterate twice here, we can just check and put the names into a set and check for both candidates directly.